controller/object_store_controller.go

```diff
@@ -0,0 +1,1150 @@
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "s3gw",
+							Image: osc.getObjectStoreImage(store),
+							Args: append([]string{
+								"--rgw-backend-store", "sfs",
+								"--rgw_frontends", fmt.Sprintf(
+									"beast port=%d, status port=%d",
+									types.ObjectStoreContainerPort,
+									types.ObjectStoreStatusContainerPort,
+								),
+							}, domainNameArgs...),
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "s3",
+									ContainerPort: types.ObjectStoreContainerPort,
+									Protocol:      "TCP",
+								},
+								{
+									Name:          "status",
+									ContainerPort: types.ObjectStoreStatusContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							EnvFrom: env,
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      genVolumeMountName(store),
+									MountPath: "/data",
+								},
+							},
+						},
+						{
+							Name:  "s3gw-ui",
+							Image: osc.getObjectStoreUIImage(store),
+							Args:  []string{},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "ui",
+									ContainerPort: types.ObjectStoreUIContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "S3GW_SERVICE_URL",
+									Value: fmt.Sprintf("http://127.0.0.1:%v", types.ObjectStoreContainerPort),
+								},
+								{
+									Name:  "S3GW_UI_PATH",
+									Value: fmt.Sprintf("/objectstore/%v", store.Name),
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: genVolumeMountName(store),
+							VolumeSource: corev1.VolumeSource{
+								PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+									ClaimName: genPVCName(store),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	registrySecretSetting, err := osc.ds.GetSetting(types.SettingNameRegistrySecret)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get registry secret setting for object store deployment")
+	}
+
+	if registrySecretSetting.Value != "" {
+		dpl.Spec.Template.Spec.ImagePullSecrets = []corev1.LocalObjectReference{
+			{
+				Name: registrySecretSetting.Value,
+			},
+		}
+	}
+
+	return osc.ds.CreateDeployment(&dpl)
+}
+
+// To avoid multiple longhorn managers acting on the same object store, only the
+// instance responsible for the longhorn volume is considered responsible for
+// the object store. This of course precludes that the volume has already been
+// created.
+func (osc *ObjectStoreController) isResponsibleFor(store *longhorn.ObjectStore) bool {
+	vol, err := osc.ds.GetVolumeRO(genPVName(store))
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		utilruntime.HandleError(fmt.Errorf("failed to find volume for object store %v: %v", store.Name, err))
+		return false
+	} else if err != nil {
+		return true
+	}
+
+	return osc.controllerID == vol.Status.OwnerID
+}
```

Comment:
Suggesting to wrap the err != nil case into a separate statement. I think this makes the code a bit safer from the nil error:
```suggestion
func (osc *ObjectStoreController) isResponsibleFor(store *longhorn.ObjectStore) bool {
	vol, err := osc.ds.GetVolumeRO(genPVName(store))
	if err != nil {
		if apierrors.IsNotFound(err) {
			return true
		}
		return false
	}
	return osc.controllerID == vol.Status.OwnerID
}
```